homeassistant/components/azure_devops/__init__.py

```diff
@@ -1,80 +1,119 @@
 from homeassistant.core import HomeAssistant
-from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryNotReady
-from homeassistant.helpers.entity import DeviceInfo, Entity
+from homeassistant.exceptions import ConfigEntryAuthFailed
+from homeassistant.helpers.entity import DeviceInfo, EntityDescription
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+    UpdateFailed,
+)
 
-from .const import CONF_ORG, CONF_PAT, CONF_PROJECT, DATA_AZURE_DEVOPS_CLIENT, DOMAIN
+from .const import CONF_ORG, CONF_PAT, CONF_PROJECT, DOMAIN
 
 _LOGGER = logging.getLogger(__name__)
 
 PLATFORMS = ["sensor"]
 
+BUILDS_QUERY: Final = "?queryOrder=queueTimeDescending&maxBuildsPerDefinition=1"
+
+
+@dataclass
+class AzureDevOpsEntityDescription(EntityDescription):
+    """Class describing Azure DevOps entities."""
+
+    organization: str = ""
+    project: DevOpsProject = None
+
 
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up Azure DevOps from a config entry."""
-    client = DevOpsClient()
-
-    try:
-        if entry.data[CONF_PAT] is not None:
-            await client.authorize(entry.data[CONF_PAT], entry.data[CONF_ORG])
-            if not client.authorized:
-                raise ConfigEntryAuthFailed(
-                    "Could not authorize with Azure DevOps. You may need to update your token"
-                )
-        await client.get_project(entry.data[CONF_ORG], entry.data[CONF_PROJECT])
-    except aiohttp.ClientError as exception:
-        _LOGGER.warning(exception)
-        raise ConfigEntryNotReady from exception
+    client: DevOpsClient = DevOpsClient()
+
+    if entry.data[CONF_PAT] is not None:
+        await client.authorize(entry.data[CONF_PAT], entry.data[CONF_ORG])
+        if not client.authorized:
+            raise ConfigEntryAuthFailed(
+                "Could not authorize with Azure DevOps. You will need to update your token"
+            )
+
+    project = await client.get_project(
+        entry.data[CONF_ORG],
+        entry.data[CONF_PROJECT],
+    )
+
+    async def async_update_data() -> list[DevOpsBuild]:
+        """Fetch data from Azure DevOps."""
+
+        try:
+            return await client.get_builds(
+                entry.data[CONF_ORG],
+                entry.data[CONF_PROJECT],
+                BUILDS_QUERY,
+            )
+        except (aiohttp.ClientError, aiohttp.ClientError) as exception:
+            raise UpdateFailed from exception
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        # Name of the data. For logging purposes.
+        name=f"{DOMAIN}_coordinator",
+        update_method=async_update_data,
+        # Polling interval. Will only be polled if there are subscribers.
+        update_interval=timedelta(seconds=300),
+    )
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = coordinator, project
```

Comment:
Move this to under `await coordinator.async_config_entry_first_refresh()`
There is no need to set this if the refresh fails